File: data-pipeline/de-normalization/src/main/java/org/ekstep/ep/samza/domain/Event.java
package org.ekstep.ep.samza.domain;


import org.ekstep.ep.samza.events.domain.Events;
import org.ekstep.ep.samza.reader.NullableValue;
import org.ekstep.ep.samza.task.DeNormalizationConfig;
import org.joda.time.DateTime;
import org.joda.time.format.DateTimeFormat;
import org.joda.time.format.DateTimeFormatter;

import java.math.BigDecimal;
import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.*;

public class Event extends Events {
    private DateTimeFormatter dateFormat = DateTimeFormat.forPattern("yyyy-MM-dd").withZoneUTC();
    private DateFormat formatter = new SimpleDateFormat("yyyy-MM-dd");

    public Event(Map<String, Object> map) {
        super(map);
    }

    public Long getEndTimestampOfDay(String date) {
        Long ets = dateFormat.parseDateTime(date).plusHours(23).plusMinutes(59).plusSeconds(59).getMillis();
        return ets;
    }

    public Long compareAndAlterEts() {
        Long eventEts = ets();
        Long currentMillis = new Date().getTime();
        String currentDate = formatter.format(currentMillis);
        Long endTsOfCurrentDate = getEndTimestampOfDay(currentDate);
        if (eventEts > endTsOfCurrentDate) telemetry.add("ets", currentMillis);
        return ets();
    }

    public Boolean isOlder(Integer periodInMonths) {
        Long eventEts = ets();
        Long periodInMillis = new DateTime().minusMonths(periodInMonths).getMillis();
        if (eventEts < periodInMillis) return true;
        else return false;
    }


    public String objectRollUpl1ID() {
        if (objectFieldsPresent() && objectRollUpl1FieldsPresent()) {
            return telemetry.<String>read("object.rollup.l1").value();
        } else return null;
    }


    public boolean objectRollUpl1FieldsPresent() {
        if (objectFieldsPresent()) {
            String objectrollUpl1 = telemetry.<String>read("object.rollup.l1").value();
            return null != objectrollUpl1 && !objectrollUpl1.isEmpty();
        } else return false;
    }

    public boolean checkObjectIdNotEqualsRollUpl1Id() {
        if (objectRollUpl1FieldsPresent() && !(objectID().equals(objectRollUpl1ID())))
            return true;
        else
            return false;
    }

    public String getKey(String type) {
        switch (type) {
            case "device":
                return did();
            case "user":
                if ("system".equalsIgnoreCase(actorType())) {
                    return "";
                } else {
                    return actorId();
                }
            case "content":
                return objectID();
            case "collection":
                return objectRollUpl1ID();
            default:
                return "";
        }
    }

    public void addMetaData(String type, Map newData) {
        switch (type) {
            case "device":
                addDeviceData(newData);
                break;
            case "user":
                addUserData(newData);
                break;
            case "content":
                addContentData(newData);
                break;
            case "dialcode":
                addDialCodeData(newData);
                break;
            case "collection":
                addCollectionData(newData);
            default:
                break;

        }
    }

    public Map addISOStateCodeToDeviceData(Map deviceData) {
        // add new statecode field
        String statecode = deviceData.get("statecode").toString();
        if (statecode != null && !statecode.isEmpty()) {
            String iso3166statecode = "IN-" + statecode;
            deviceData.put("iso3166statecode", iso3166statecode);
            return deviceData;
        } else return deviceData;
    }

    public void addDeviceData(Map newData) {
        NullableValue<Map<String, Object>> previousData = telemetry.read(path.deviceData());
        Map<String, Object> deviceData = previousData.isNull() ? new HashMap<>() : previousData.value();
        // add new statecode field
        if (!deviceData.isEmpty()) {
            deviceData = addISOStateCodeToDeviceData(deviceData);
        }
        deviceData.putAll(newData);
        telemetry.add(path.deviceData(), deviceData);
        setFlag(DeNormalizationConfig.getDeviceLocationJobFlag(), true);
    }

    public void addUserData(Map newData) {
        NullableValue<Map<String, Object>> previousData = telemetry.read(path.userData());
        Map<String, Object> userdata = previousData.isNull() ? new HashMap<>() : previousData.value();
        userdata.putAll(newData);
        telemetry.add(path.userData(), userdata);
        if (newData.size() > 2)
            setFlag(DeNormalizationConfig.getUserLocationJobFlag(), true);
        else
            setFlag(DeNormalizationConfig.getUserLocationJobFlag(), false);
    }

    public void addContentData(Map newData) {
        NullableValue<Map<String, Object>> previousData = telemetry.read(path.contentData());
        Map<String, Object> contentData = previousData.isNull() ? new HashMap<>() : previousData.value();
        contentData.putAll(newData);
        telemetry.add(path.contentData(), contentData);
        setFlag(DeNormalizationConfig.getContentLocationJobFlag(), true);
    }

    public void addCollectionData(Map newData) {
        NullableValue<Map<String, Object>> previousData = telemetry.read(path.collectionData());
        Map<String, Object> collectionData = previousData.isNull() ? new HashMap<>() : previousData.value();
        collectionData.putAll(newData);
        telemetry.add(path.collectionData(), collectionData);
        setFlag(DeNormalizationConfig.getCollectionLocationJobFlag(), true);
    }

    public void addDialCodeData(Map newData) {
        telemetry.add(path.dialCodeData(), newData);
        setFlag(DeNormalizationConfig.getDialCodeLocationJobFlag(), true);
    }

    public void setFlag(String key, Object value) {
        NullableValue<Map<String, Object>> telemetryFlag = telemetry.read(path.flags());
        Map<String, Object> flags = telemetryFlag.isNull() ? new HashMap<>() : telemetryFlag.value();
        if (!key.isEmpty()) flags.put(key, value);
        telemetry.add(path.flags(), flags);
    }

    public Map<String, Object> getFlag() {
        NullableValue<Map<String, Object>> telemetryFlag = telemetry.read(path.flags());
        Map<String, Object> flags = telemetryFlag.isNull() ? new HashMap<>() : telemetryFlag.value();
        return flags;
    }

    public String getUpgradedVersion() {
        BigDecimal updatedVer = BigDecimal.valueOf(Double.parseDouble(telemetry.read(path.ver()).value().toString())).add(BigDecimal.valueOf(0.1));
        return updatedVer.toString();
    }

    public void markFailure(String error, DeNormalizationConfig config) {
        telemetry.addFieldIfAbsent("flags", new HashMap<String, Boolean>());
        telemetry.add("flags.denorm_processed", false);

        telemetry.addFieldIfAbsent("metadata", new HashMap<String, Object>());
        telemetry.add("metadata.denorm_error", error);
        telemetry.add("metadata.src", config.jobName());
    }


}

File: data-pipeline-flink/de-normalization/src/main/scala/org/sunbird/dp/domain/Event.scala
package org.sunbird.dp.domain

import java.util

import org.apache.commons.lang3.StringUtils
import org.joda.time.format.DateTimeFormat
import org.sunbird.dp.task.DenormalizationConfig
import org.joda.time.DateTime
import org.joda.time.LocalDate
import org.sunbird.dp.reader.NullableValue
import org.joda.time.format.DateTimeFormatter
import scala.collection.mutable.Map
import collection.JavaConverters._

class Event(eventMap: util.Map[String, AnyRef], partition: Integer) extends Events(eventMap, partition) {

  private[this] val df = DateTimeFormat.forPattern("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'").withZoneUTC
  private[this] val df2 = DateTimeFormat.forPattern("yyyy-MM-dd'T'HH:mm:ss.SSS").withZoneUTC();
  private val jobName = "PipelinePreprocessor"

  def addDeviceProfile(deviceProfile: DeviceProfile): Unit = {

    val deviceMap = new util.HashMap[String, Object]();
    val userDeclaredMap = new util.HashMap[String, Object]();
    val iso3166statecode = addISOStateCodeToDeviceProfile(deviceProfile);
    val ldata = Map[String, AnyRef]("countrycode" -> deviceProfile.countryCode,
      "country" -> deviceProfile.country,
      "statecode" -> deviceProfile.stateCode,
      "state" -> deviceProfile.state,
      "city" -> deviceProfile.city,
      "statecustomcode" -> deviceProfile.stateCodeCustom,
      "statecustomname" -> deviceProfile.stateCustomName,
      "districtcustom" -> deviceProfile.districtCustom,
      "devicespec" -> deviceProfile.devicespec,
      "firstaccess" -> deviceProfile.firstaccess.asInstanceOf[AnyRef],
      "iso3166statecode" -> iso3166statecode)
    
    deviceMap.putAll(ldata.asJava);
    userDeclaredMap.putAll(Map[String, String]("state" -> deviceProfile.userDeclaredState, "district" -> deviceProfile.userDeclaredDistrict).asJava);
    deviceMap.put("userdeclared", userDeclaredMap)
    telemetry.add(Path.DEVICE_DATA, deviceMap);
    setFlag("device_denorm", true);
  }

  def getUserProfileLocation(): Option[(String, String, String)] = {

    val userData: NullableValue[util.Map[String, AnyRef]] = telemetry.read(Path.USER_DATA);
    if (!userData.isNull())
      Some(userData.value().get("state").asInstanceOf[String], userData.value().get("district").asInstanceOf[String], "user-profile")
    else
      None
  }

  def getUserDeclaredLocation(): Option[(String, String, String)] = {

    val deviceData: NullableValue[util.Map[String, AnyRef]] = telemetry.read(Path.DEVICE_DATA);
    if (!deviceData.isNull() && null != deviceData.value().get("userdeclared")) {
      val userDeclared = deviceData.value().get("userdeclared").asInstanceOf[util.Map[String, String]]
      Some(userDeclared.get("state"), userDeclared.get("district"), "user-declared")
    } else
      None
  }

  def getIpLocation(): Option[(String, String, String)] = {
    val deviceData: NullableValue[util.Map[String, AnyRef]] = telemetry.read(Path.DEVICE_DATA);
    if (!deviceData.isNull())
      Some(deviceData.value().get("state").asInstanceOf[String], deviceData.value().get("districtcustom").asInstanceOf[String], "ip-resolved")
    else
      None
  }

  def addDerivedLocation(derivedData: (String, String, String)) {
    val locMap = new util.HashMap[String, String]();
    locMap.put(Path.STATE_KEY, derivedData._1)
    locMap.put(Path.DISTRICT_KEY, derivedData._2)
    locMap.put(Path.DERIVED_LOC_FROM, derivedData._3)
    telemetry.add(Path.DERIVED_LOC_DATA, locMap);
    setFlag("loc_denorm", true);
  }

  def compareAndAlterEts(): Long = {
    val eventEts = ets();
    val endTsOfCurrentDate = DateTime.now().plusDays(1).withTimeAtStartOfDay().minusMillis(1).getMillis;
    if (eventEts > endTsOfCurrentDate) telemetry.add("ets", endTsOfCurrentDate);
    ets();
  }

  def isOlder(periodInMonths: Int): Boolean = {
    val eventEts = ets();
    val periodInMillis = new DateTime().minusMonths(periodInMonths).getMillis();
    eventEts < periodInMillis
  }

  def objectRollUpl1ID(): String = {
    telemetry.read[String]("object.rollup.l1").value();
  }

  def objectRollUpl1FieldsPresent(): Boolean = {

    val objectrollUpl1 = telemetry.read[String]("object.rollup.l1").value();
    null != objectrollUpl1 && !objectrollUpl1.isEmpty();
  }

  def checkObjectIdNotEqualsRollUpl1Id(): Boolean = {
    objectRollUpl1FieldsPresent() && !(objectID().equals(objectRollUpl1ID()))
  }

  def addUserData(newData: Map[String, AnyRef]) {
    val previousData: NullableValue[util.Map[String, AnyRef]] = telemetry.read(Path.USER_DATA);
    val userdata = if (previousData.isNull()) new util.HashMap[String, AnyRef]() else previousData.value();
    userdata.putAll(newData.asJava);
    telemetry.add(Path.USER_DATA, userdata);
    if (newData.size > 2)
      setFlag("user_denorm", true);
    else
      setFlag("user_denorm", false);
  }

  def addContentData(newData: Map[String, AnyRef]) {
    val convertedData = getEpochConvertedContentDataMap(newData);
    val previousData: NullableValue[util.Map[String, AnyRef]] = telemetry.read(Path.CONTENT_DATA);
    val contentData = if (previousData.isNull()) new util.HashMap[String, AnyRef]() else previousData.value();
    contentData.putAll(convertedData.asJava);
    telemetry.add(Path.CONTENT_DATA, contentData);
    setFlag("content_denorm", true);
  }

  def addCollectionData(newData: Map[String, AnyRef]) {
    val collectionMap = new util.HashMap[String, AnyRef]();
    val convertedData = getEpochConvertedContentDataMap(newData);
    collectionMap.putAll(convertedData.asJava);
    telemetry.add(Path.COLLECTION_DATA, collectionMap);
    setFlag("coll_denorm", true);
  }

  def getEpochConvertedContentDataMap(data: Map[String, AnyRef]): Map[String, AnyRef] = {

    val lastSubmittedOn = data.get("lastsubmittedon");
    val lastUpdatedOn = data.get("lastupdatedon");
    val lastPublishedOn = data.get("lastpublishedon");
    if (lastSubmittedOn.nonEmpty && lastSubmittedOn.get.isInstanceOf[String]) {
      data.put("lastsubmittedon", getConvertedTimestamp(lastSubmittedOn.get.asInstanceOf[String]).asInstanceOf[AnyRef]);
    }
    if (lastUpdatedOn.nonEmpty && lastUpdatedOn.get.isInstanceOf[String]) {
      data.put("lastupdatedon", getConvertedTimestamp(lastUpdatedOn.get.asInstanceOf[String]).asInstanceOf[AnyRef]);
    }
    if (lastPublishedOn.nonEmpty && lastPublishedOn.get.isInstanceOf[String]) {
      data.put("lastpublishedon", getConvertedTimestamp(lastPublishedOn.get.asInstanceOf[String]).asInstanceOf[AnyRef]);
    }
    data;
  }

  def addDialCodeData(newData: Map[String, AnyRef]) {
    val dialcodeMap = new util.HashMap[String, AnyRef]();
    dialcodeMap.putAll(getEpochConvertedDialcodeDataMap(newData).asJava);
    telemetry.add(Path.DIALCODE_DATA, dialcodeMap);
    setFlag("dialcode_denorm", true);
  }

  private def getEpochConvertedDialcodeDataMap(data: Map[String, AnyRef]): Map[String, AnyRef] = {

    val generatedOn = data.get("generatedon");
    val publishedOn = data.get("publishedon");
    if (generatedOn.nonEmpty && generatedOn.get.isInstanceOf[String]) {
      data.put("generatedon", getConvertedTimestamp(generatedOn.get.asInstanceOf[String]).asInstanceOf[AnyRef])
    }
    if (publishedOn.nonEmpty && publishedOn.get.isInstanceOf[String]) {
      data.put("publishedon", getConvertedTimestamp(publishedOn.get.asInstanceOf[String]).asInstanceOf[AnyRef]);
    }
    data
  }

  def getTimestamp(ts: String, df: DateTimeFormatter): Long = {
    try {
      df.parseDateTime(ts).getMillis();
    } catch {
      case ex: Exception =>
        0L;
    }
  }

  def getConvertedTimestamp(ts: String): Long = {
    val epochTs = getTimestamp(ts, df);
    if (epochTs == 0) {
      getTimestamp(ts, df2);
    } else {
      epochTs;
    }
  }

  def setFlag(key: String, value: Boolean) {
    val telemetryFlag: NullableValue[util.Map[String, AnyRef]] = telemetry.read(Path.FLAGS);
    val flags = if (telemetryFlag.isNull()) new util.HashMap[String, AnyRef]() else telemetryFlag.value()
    flags.put(key, value.asInstanceOf[AnyRef]);
    telemetry.add(Path.FLAGS, flags);
  }

  def addISOStateCodeToDeviceProfile(deviceProfile: DeviceProfile): String = {
    // add new statecode field
    val statecode = deviceProfile.stateCode;
    if (statecode != null && !statecode.isEmpty()) {
      return "IN-" + statecode;
    } else return "";
  }

}
